torch/distributions/transforms.py

```diff
@@ -535,6 +536,83 @@ def log_abs_det_jacobian(self, x, y):
         return result.expand(shape)
 
 
+class CorrCholeskyTransform(Transform):
+    r"""
+    Transforms an uncontrained real vector :math:`x` with length :math:`D*(D-1)/2` into the
+    Cholesky factor of a D-dimension correlation matrix. This Cholesky factor is a lower
+    triangular matrix with positive diagonals and unit Euclidean norm for each row.
+    The transform is processed as follows:
+
+        1. First we convert x into a lower triangular matrix in row order.
+        2. For each row :math:`X_i` of the lower triangular part, we apply a *signed* version of
+        class :class:`StickBreakingTransform` to transform :math:`X_i` into a
+        unit Euclidean length vector using the following steps:
+            a. Scales into the interval :math:`(-1, 1)` domain: :math:`r_i = \tanh(X_i)`.
+            b. Transforms into an unsigned domain: :math:`z_i = r_i^2`.
+            c. Applies :math:`s_i = StickBreakingTransform(z_i)`.
+            d. Transforms back into signed domain: :math:`y_i = sign(r_i) * \sqrt{s_i}`.
+    """
+    domain = constraints.real_vector
+    codomain = constraints.corr_cholesky
+    bijective = True
+    # Note that since we add a dim in _call this should actually be 2 for the inverse transform.
+    event_dim = 1
+
+    def _call(self, x):
+        x = torch.tanh(x)
+        eps = torch.finfo(x.dtype).eps
+        x = x.clamp(min=-1 + eps, max=1 - eps)
+        n = (1 + math.sqrt(1 + 8 * x.shape[-1])) / 2
+        if round(n) - n > eps:
+            raise ValueError(f'The size of last dimension is {y.shape[-1]} which cannot be expressed ' +
+                             '(D * (D-1)) / 2, i.e. the lower triangular part of a square D x D matrix.')
+        n = round(n)
+        r = torch.zeros(x.shape[:-1] + (n, n))
+        # TODO: Explore a JIT friendly way to convert a tensor to
+        # a lower triangular matrix.
+        tril_idx = torch.tril_indices(n, n, -1, device=x.device)
```

Comment:
Note that `tril_indices` isn't JIT friendly since n cannot be a pytorch scalar, so to make this generalize to arbitrary `n` and work with JIT, we'll need to use something else here.

Reply:
I guess we can make two branches: 1 under JIT and one not. In JIT tracing, we can do
```
tril_idx = (r + 1).tril(diagonal=-1) > 0.5  # I'm not sure if this works under JIT...
```

Reply:
Thanks. Your comment on using masked indexing seems to work with JIT.